### PR TITLE
[Partitioner] Polish code and fix a few bugs to make optimization sequence consistent for each run.

### PR DESCRIPTION
**Pull request by beicy**
*Description*:
This PR 1) Make the nodes/functions working sequence fixed based on their names. Therefore, the internal optimization sequence is consistent for each run.
2) Fixed a few bugs related to communication cost.
3) Polished partitions combination optimization.

*Testing*:
ninja check, runtime-resnet.

*Documentation*:
#2298 
[Optional Fixes #issue]

Please see a detailed explanation of how to fill out the fields in the relevant sections in PULL_REQUEST.md.

### CODE REVIEW

File: include/glow/Partitioner/Partitioner.h [hunk 1]
```diff
@@ -29,12 +29,21 @@ using ComputeTimeMapTy = std::unordered_map<Node *, float>;
...
-/// a set of nodes sets.
+/// newly-created functions;
 using NodeToFunctionMapTy = llvm::DenseMap<Node *, Function *>;
-using FunctionToNodesMapTy = llvm::DenseMap<Function *, NodesSetTy>;
+
+/// A mapping of newly-created functions along with a set of nodes sets. The
+/// overloaded compare function to make sure the map is sorted by the key's
+/// name(i.e. the function's name) which makes the optimization sequence is
+/// consistent for each run.
+struct ptr_less {
```
Comment (opti-mix) [inline include/glow/Partitioner/Partitioner.h]: First, the name `ptr_less` does not follow our usual CamelCase naming convention for types.

And then, it actually compares function names. So, may be it should be called `FunctionNameComparator`?

File: include/glow/Partitioner/PartitionerUtils.h [hunk 2]
```diff
@@ -33,6 +33,11 @@ struct GraphMemInfo {
   uint64_t constMemSize;
 
   GraphMemInfo() : inMemSize(0), outMemSize(0), constMemSize(0){};
+
+  // Get the total memory size of each partition.
+  const uint64_t getTotalMemSize() {
```
Comment (rdzhabarov) [inline include/glow/Partitioner/PartitionerUtils.h]: I meant `const` method. like 

uint64_t getTotalMemSize() const {
...
}

File: include/glow/Partitioner/PartitionerUtils.h [hunk 3]
```diff
@@ -50,7 +55,12 @@ std::vector<Node *> getOutUsers(const std::set<Node *> &nodes);
 std::vector<Node *>
 getOutUsersWithOnePredecessor(const std::set<Node *> &nodes);
 
+/// Return the memory usage of the output caused by \p node who has users not in
+/// the set \p nodes.
+uint64_t getOutMemPerNode(const std::set<Node *> &nodes, const Node *node);
```
Comment (rdzhabarov) [inline include/glow/Partitioner/PartitionerUtils.h]: this could also be made `const`.

File: lib/Partitioner/Partitioner.cpp [hunk 4]
```diff
@@ -201,46 +201,60 @@ void Partitioner::initOpComputeTime() {
...
+        }
+      }
+      if (flag) {
+        // This partition only has one successor.
+        Function *cur = (*it).first;
+        Function *suc = partitions[outUsers[0]];
+        NodesSetTy tmp = (nodesSet.find(suc))->second;
+        GraphMemInfo cost1 = partitions.getGraphMemInfo(cur);
+        GraphMemInfo cost2 = partitions.getGraphMemInfo(suc);
+        if (cost1.getTotalMemSize() + cost2.getTotalMemSize() -
```
Comment (rdzhabarov) [inline lib/Partitioner/Partitioner.cpp]: not very readable.

**Pull request merged**